### ISSUE TITLE
ManagedIdentityCredential treats all HTTP status codes as a TextError

### ISSUE THREAD

**Issue by thovoll** [Azure/azure-sdk-for-rust#428]
https://github.com/Azure/azure-sdk-for-rust/blob/b5137482b420627579a6382dd9a2490ca17868e1/sdk/identity/src/token_credentials/managed_identity_credentials.rs#L65

Should we be more accurate and specific about non-successful results?

**Comment by JoshGendein** [comment]
I can start working on this if no one has picked it up yet.

**Comment by thovoll** [comment]
Thanks @JoshGendein, go for it!

**Comment by thovoll** [comment]
Also note #489

**Comment by JoshGendein** [comment]
@thovoll How have you been running/testing the managed identity credential code locally?

**Comment by thovoll** [comment]
@JoshGendein I have an [IMDS endpoint](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http) simulator that runs on my dev box and responds to http://169.254.169.254/metadata/identity/oauth2/token. `ManagedIdentityCredential` calls the simulator, which in turn calls AAD with a pre-configured identity (and creds) to get a token and pass it back to the Rust SDK AMI code. That way I can actually use the token to make calls to the Azure resource, which is not strictly necessary for testing `ManagedIdentityCredential` because the actual bearer token is opaque anyway. What matters is that the simulator returns a properly formatted JSON response as [documented here](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http):
```
HTTP/1.1 200 OK
Content-Type: application/json
{
  "access_token": "eyJ0eXAi...",
  "refresh_token": "",
  "expires_in": "3599",
  "expires_on": "1506484173",
  "not_before": "1506480273",
  "resource": "https://management.azure.com/",
  "token_type": "Bearer"
}
```

You should also be able to have the local simulator respond to another local endpoint that you can tell `ManagedIdentityCredential` about via the [`MSI_ENDPOINT_ENV_KEY`](https://github.com/Azure/azure-sdk-for-rust/blob/b5137482b420627579a6382dd9a2490ca17868e1/sdk/identity/src/token_credentials/managed_identity_credentials.rs#L44) environment variable.

This actually seems very similar to what the mock testing framework in the Azure Rust SDK is doing by replaying API responses while running the test. Maybe there's a way to us that (@rylev @MindFlavor).

A real integration test would involve running the code in a place that actually supports AMI, like an Azure VM.

**Comment by thovoll** [comment]
Also note #420

**Comment by cataggar** [comment]
Fixed by #511.
